### PR TITLE
BUG - get_method/property uses "base" rather than "most derived"

### PR DESCRIPTION
**Pull request by dand-oss**
If you register the same function name in a derived class and in a base class
and call the function via a derived instance

C++ calls the derived method
RTTR calls the base method

RTTR searches from the "base class" to the derived class for the matching name.

The linear search of prop/meth vecs must use reverse iterator

staring from the most derived class registered.

### CODE REVIEW

File: src/rttr/type.cpp [hunk 1]
```diff
@@ -216,12 +216,12 @@ bool type::destroy(variant& obj) const RTTR_NOEXCEPT
 property type::get_property(string_view name) const RTTR_NOEXCEPT
 {
     const auto& vec = get_raw_type().m_type_data->get_class_data().m_properties;
-    auto ret = std::find_if(vec.cbegin(), vec.cend(),
+    auto ret = std::find_if(vec.crbegin(), vec.crend(),
```
Comment (acki-m) [inline src/rttr/type.cpp]: please add a comment, why you are using the reverse iterator

**Comment by acki-m** [comment]
Can you please add a unit test for both scenarios?

File: src/unit_tests/method/method_order_test.cpp [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    std::string whoami() { return "I am a base method" ; }
+    virtual std::string vwhoami() { return "I am a virtual base method" ; }
+
+    RTTR_ENABLE()
+
+};
+
+
+// RTTR_ENABLE incompatible with clang
+#define RTTR_ENABLE_OVERRIDE(...) \
```
Comment (acki-m) [inline src/unit_tests/method/method_order_test.cpp]: What is the problem here? Why is this necessary?

File: src/unit_tests/method/method_order_test.cpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        const std::string& meth_name) {
+
+    const auto& inst_t = type::get<TT>();
+
+    // find method
+    const auto& meth = inst_t.get_method(meth_name);
+    REQUIRE(meth.is_valid() == true);
+
+    // invoke via method
+    variant&& iam_var = meth.invoke(inst);
```
Comment (acki-m) [inline src/unit_tests/method/method_order_test.cpp]: Why `&&` ?
Should be enough:
`auto iam_var = meth.invoke(inst);`

File: src/unit_tests/method/method_order_test.cpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        ;
+    registration::class_<method_order_test_derived>("method_order_test_derived")
+        .method("whoami", &method_order_test_derived::whoami)
+        .method("vwhoami", &method_order_test_derived::vwhoami)
+        ;
+}
+
+// approach 1 looks up the "method instance" without considering argument type and invoke the registered function()
+template<class TT>
+std::string rttr_invoke_approach1(
```
Comment (acki-m) [inline src/unit_tests/method/method_order_test.cpp]: No need for a function template. 
Try this:
```
std::string rttr_invoke_approach1(rttr::instance inst, rttr::string_view meth_name);
```

PS: All braces, `{`, should be placed in the same column, otherwise the CI tests will not pass. Thats the coding style for RTTR. Check travis, first build.

File: src/rttr/type.cpp [hunk 5]
```diff
@@ -309,8 +315,11 @@ method type::get_method(string_view name) const RTTR_NOEXCEPT
 method type::get_method(string_view name, const std::vector<type>& type_list) const RTTR_NOEXCEPT
 {
     const auto raw_t = get_raw_type();
-    for (const auto& meth : raw_t.m_type_data->get_class_data().m_methods)
+
+    const auto& methvec = raw_t.m_type_data->get_class_data().m_methods;
+    for (auto mit = methvec.crbegin() ; mit != methvec.crend() ; mit++)
```
Comment (acki-m) [inline src/rttr/type.cpp]: better use prefix operator:
`for (auto mit = methvec.crbegin() ; mit != methvec.crend() ; ++mit)`

File: src/unit_tests/method/method_order_test.cpp (hunk 3 above)
Comment (dand-oss) [reply src/unit_tests/method/method_order_test.cpp]: References are better than variables because they ensure no copies are made.

Best
1.  const ref& - ensures reference cannot be modified.  When you see this you know what you have.
2.  ref&& - universal reference allows modification as lvalue
3.  variable - requires a copy
Worst

File: src/unit_tests/method/method_order_test.cpp (hunk 2 above)
Comment (dand-oss) [reply src/unit_tests/method/method_order_test.cpp]: Use RTTR_ENABLE instead of RTTR_OVERRIDE and the CI will fail.

Clang in your CI fails if RTTR_ENABLE is used twice in a single file.  The reason is NOT using "override" on the virtual is an error rather than a warning.

File: src/rttr/type.cpp (hunk 5 above)
Comment (dand-oss) [reply src/rttr/type.cpp]: done

File: src/unit_tests/method/method_order_test.cpp (hunk 3 above)
Comment (dand-oss) [reply src/unit_tests/method/method_order_test.cpp]: Done - const auto& works

File: src/unit_tests/method/method_order_test.cpp (hunk 4 above)
Comment (dand-oss) [reply src/unit_tests/method/method_order_test.cpp]: done - prefer const rttr::instance& to templaet

File: src/unit_tests/method/method_order_test.cpp (hunk 3 above)
Comment (acki-m) [reply src/unit_tests/method/method_order_test.cpp]: Please see here:
https://stackoverflow.com/questions/4986673/c11-rvalues-and-move-semantics-confusion-return-statement
The variant has a move-ctor. So just use `auto` and you are fine.

File: src/unit_tests/method/method_order_test.cpp (hunk 2 above)
Comment (acki-m) [reply src/unit_tests/method/method_order_test.cpp]: I have never encountered such a warning, How do you get this?
Anyway, this should be fixed globally not locally.
I could disable this clang warning, because it is in the current case wrong.

**Comment by acki-m** [comment]
@dand-oss 
any updates here? I would also change something in this PR...

**Comment by dand-oss** [comment]
@acki-m 
My understanding at this point
- all questions answered
- all requests implemented
- all unit test pass
- rebased to latest upstream

Comment thread nesting is a bit confusing
If there is anything else you'd like me to do, let me know

**Comment by acki-m** [comment]
@dand-oss 
Did you read my latest comment in the modified code sections?

The biggest issue is this the usage of `RTTR_ENABLE_OVERRIDE`
You need to tell me how I can get this warning myself.

**Comment by acki-m** [comment]
I need at least the info to reproduce the warning, otherwise I cannot fix the source of the issue.

**Comment by acki-m** [comment]
@dand-oss 
Please give me an update here.

**Comment by dand-oss** [comment]
> > I need at least the info to reproduce the warning, otherwise I cannot fix the source of the issue.
> Please give me an update here.

I created a branch to demonstrate the clang problem.

Clang override error
https://github.com/rttrorg/rttr/pull/189

struct compile_base
{
    RTTR_ENABLE()
};

struct compile_derived : compile_base
{
    // Clang options want "override" on virtuals inherited from base
    RTTR_ENABLE(compile_base)
};

/home/travis/build/rttrorg/rttr/src/unit_tests/misc/test_misc.cpp:57:5: error: 'get_derived_info' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]
/home/travis/build/rttrorg/rttr/src/unit_tests/misc/test_misc.cpp:57:5: error: 'get_ptr' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]
/home/travis/build/rttrorg/rttr/src/unit_tests/misc/test_misc.cpp:57:5: error: 'get_type' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]

**Comment by acki-m** [comment]
@dand-oss 
Please update your PR with the newest changes from `master`, remove your workaround and then we can finally merge. 👍

**Comment by dand-oss** [comment]
rebased, squashed, and test passes

**Comment by acki-m** [comment]
@dand-oss 
no need to squash the commit, I can do this too.
However, there is still a `RTTR_ENABLE_OVERRIDE` in the sources. Can you please remove it?
Then I can merge it.
